fixes for various navigation / accessibility issues

MarcSabatella opened the pull request:
See https://musescore.org/en/node/282091.  The navigation code is full of special cases to control the order of navigation through the object hierarchy.  The section I fixed was trying to find the note, rest, or chord that is the parent of a selected element, but it only checked a couple of different element types that have notes, rests, or chords as parents.  Rather than add cases for the missed types (fingering, tremolo, arpeggios, symbols, etc), I check the parent type directly.

This code, BTW, was developed at more or less the same time as some major refactoring in the element typing scheme (use of isNote() etc) and some changes in the object hierarchy as well, so there are some inconsistencies throughout and opportunities to clean things up.  Plus we need to address some of the TODO's left when some of those refactorings were performed.  For example, Alt+left/right no longer catches articulations or lyrics because the code in ChordRest::nextArticulationOrLyrics(0 is disabled.

So while an overhaul of this code in general would be very welcome especially for accessibility, right now I'm just focused on this one piece of the puzzle because of the detrimental effect it has on "fingering mode" as per the issue linked above.

MarcSabatella commented:
Although my original reason for working on this today was just about entering fingerings, the underlying and related accessibility regressions have been bothering me for some time, so I went ahead and fixed the navigation for articulations & lyrics, the screenreader feedback for articulations, and the loss selection after deleting a fingering or other element attached to a note or rest.

MarcSabatella commented:
Added one more commit, to restore a lost selection upon using alt+left/right.  I don't actually restore the selected element, too much danger of it no longer being valid.  Instead I simply keep track of the tick and track of the selected element (updating it on each change in Selection::updateState()).  Then if the navigation commands are used with no selection, I try to find a valid note or rest to select first.  If no previous selection has been made at all, I continue to start at the very beginning of the score (eg, initial system barline or clef).

In addition to protecting the user against losing selection permanently by pressing Esc (it's still lost as it should be, it's just possible to regain it), this is good insurance against any other operations that may cause unintended loss of selection.  There are still a number of these, but I will not try to address them here, I'd rather see this PR merged as is before we worry too much about other commands.

RobFog commented:
Typo: "mutiple" needs another "l".

MarcSabatella commented:
(fixed typos)

MarcSabatella commented:
An additional commit to allow alt+left/right to navigate frames (and make the screenreader read the text content).  Also fixes a couple of crashes, hangs, and other glitches that occur otherwise.

MarcSabatella commented:
Note: also fixes https://musescore.org/en/node/292928

anatoly-os commented on libmscore/note.cpp:
```diff
@@ -2898,7 +2898,7 @@ Element* Note::nextElement()
                   if (next)
                         return next;
-                  else if (_tieFor)
+                  else if (_tieFor && !_tieFor->segmentsEmpty())
```
Isn't it better to create standalone method with a clear name which incapsulates that check?

MarcSabatella replied:
You mean, to check for both the existance of the tieFor and the validity of it in one call?  I can do that, will add another commit.

anatoly-os commented on libmscore/input.cpp:
```diff
@@ -60,7 +60,7 @@ Fraction InputState::tick() const
 ChordRest* InputState::cr() const
       {
-      return _segment ? toChordRest(_segment->element(_track)) : 0;
+      return _segment && _track >= 0 ? toChordRest(_segment->element(_track)) : 0;
```
That `_track >= 0` requires additional comments in the code, imho. Please, mention in the comments, why you decided to change the semantic of the pretty fundamental method?

MarcSabatella replied:
OK.  Pretty sure there was no reason other than protection against a crash if _track ends up being -1 at the point where this called, but I'll review.

MarcSabatella replied:
After investigation, I confirm the possibility that navigation through frames or other elements not attached to any particular staff can leave track at -1, so the above is needed to avoid a potential crash.  It has no other effect, in all other cases track is >= 0.  But I've added a comment to that effect in a commit I'll push soon.

anatoly-os commented:
@MarcSabatella thank you for the improvements you made in this PR, very much appreciated! I left some comments as a review and want to sum up my thoughts:

1. I clearly understand the changes made in 713d9a1e6bccd4a2fc60c1013bd65ac2ea616999, eabe6bc3a109ac347de726d77708ad63e8541fae, 2e212ce3059e37d73110f123100f6acfdf4fc639 and ready to merge them. Still, I think of the script test which can actually verify the changes. Imagine a score with the elements, then using the accessibility navigation keys you reach the element you couldn't reach before and remove it. The test will pass with your changes and will fail without the fixes. Could you please add such tests for each commit?
2. Regarding the changes made in b7f534b626fbbaf28f5b3885dbc59df304361d55, I didn't get why `cmdDeleteSelection` method is responsible for actually restoring the selection. The code needs revising.
3. Regarding the 1cffcdcd64c62b842d8981c69b1836b0a23d2d15, I left a comment in the issue https://musescore.org/en/node/292269#comment-939300
4. The fix made in 52e52be821fa48ca198a0bf5cd6d8ea48570d868 looks good, but doesn't relate to other changes made in this PR (which are mainly related to Accessibility Navigation shortcuts). There is no need to put everything in one PR unless we want to discuss each commit in big comments like the current one :)
5. The changes made in df0fccdec7ba689eb3bc3f80c682701043225867 and 0c34ee08979dc8e187ced13b1bf0025e8ee29f1a are unclear to me neither the motivation nor the implementation. I would open a separate PR to discuss things there.
6. Same here ca298da3035403bc738c7a12d08cd7d364bb544f
7. I get the motivation in a3104295a549ba585a5a68142799c5561a928c1e, but the changes we make in the palettes need additional discussions due to the scheduled palettes refinement.
8. Regarding the a3104295a549ba585a5a68142799c5561a928c1e, what is the motivation?

MarcSabatella commented:
Thanks for the review!

1. I agree script tests can be made for the navigation changes, and I will add some.  I have had trouble getting script tests to pass - trivial differences I end up having to edit by hand - but maybe someone can help walk me through this on Telegram.
2. I didn't make any changes to the overall structure of how selection happens after deletion, just fixed the cases that were broken in the existing code.  I agree someday it would be good to revisit the general approach.
3. Commented in the issue.  @shoogle feel free to also give feedback the ability to regain a lost selection.
4. The crash *is* in the next-element and previous-element commands.  As for why I've combined everything into one PR, that would not have been my first choice either.  But I need a build that includes all these fixes in order for my blind student to be using right now as we prepare for the start of classes in a few days.  I don't like monster PR's like this either, which is why I have been continually bringing this PR up on Telegram, so the earlier work could get merged into master and allow me to start new PR's that build on it.
5. I will open issues to explain the cases.  The short version is, a blind user otherwise will have no way of ready the text in frames, such as titles of scores, or instructions on homework assignment placed in text frames.  Again, I need this all working within the next few days in order for my bind student and I to be ready for the start of school, so I need to be able to build one branch that contains ll the necessary changes, or I'd have made it a separate PR.
6. This came up in my accessibility review with my student, but I will open an issue for this too.  Short version: the screenreader report right now is broken, it read key signatures like "F# major" as "F major", because screenreaders don't necessarily know how to pronounce Unicode characters.  So we need to use (translated) English words in the screenreader info even while we use the Unicode characters for the status bar or tooltips.
7. Understood.  Again, I need this working now, so I did the work, but these changes are self-contained so I can add another commit to revert this if you like, then after merging the rest of the PR (hopefully!), I can create a new branch containing only that change and use my own build from that branch with my student.
8. Typo, that's the same as the commit for 7?  Maybe you meant https://github.com/musescore/MuseScore/pull/5221/commits/f3fc8a18db03dd2f91dc50e6f3f5d3831ec26a5c?  This also is something that commonly comes up in accessibility reviews, including mine with my student the other day, but Peter has had on his todo list from RNIB for a long time as well.  I will add an issue.  Or if you meant something else, let me know.

MarcSabatella commented:
- added https://musescore.org/en/node/293298 and https://musescore.org/en/node/293300 for changes not already linked to issues
- @shoogle confirmed via Telegram that my palette changes do not create conflicts for @AnandHemachandran, so I prefer to leave these here
- working on a new commit for the code changes suggested
- will add tests where possible (mostly for navigation)

MarcSabatella commented:
Added some comments, refactored where requestion, added links to issues where missing.

Still todo: adding script tests, which I will get to later tonight.  Out of town and very busy next couple of days and may or may not have much internet access, so really hoping to get everything I need to taken care of today.

MarcSabatella commented:
Script test added that covers the navigation changes.  No way I know to do automated tests for the screenreader changes, but we'll be giving this pretty extensive usage right away.  While there are still a few open accessibility issues on my radar, none of them are critical for me right now to start teaching, so I don't plan to add more commits to this PR.

MarcSabatella commented:
As promised, no new commits :-), but I did go ahead and expand the commit messages for each of the main commits.  To be honest, I've been submitting PR's for seven years now (!) and never actually realized one could create multi-line commit messages.  So I've always used the PR itself, or the issue report, for comments.  Now that I know how to do it, I'll try to get in the habit.

anatoly-os commented:
I created #5265, #5266 and #5267 for the changes I completely understand, agree and ready to merge. Let's see the build results.

I need more time to dive into the changes from 8eda3a18fd90003a1676bd444ef0500d42d1eb5e, 7ec38af6ef75bc13cf2a49d77e1dac27f05b1333, 4337093dfd5b47a90a4b50a1347a43eb12f9b646, and f99ede2fc67926749391f3795ad0b2f74fec0940.

anatoly-os commented:
@MarcSabatella I merged the changes from 624d38a2396e9bb5b70199e0a5edb14e1f0ddfd2, d7a659e47e06eb187c095cddf36e0897fbde90d8, ec5cb87db0f277b76d0ea9f8d6ed8a4272e823f4, 779d64795636c23d3ae36d854267a10987022137 and 6634810a232c4667992f612b68e2c36e83420d2e to master. Also, I merged the changes from 38baf32dd5d28bdd3ac22046716747147de975dd related to the 6634810a232c4667992f612b68e2c36e83420d2e.

Could you please remove those commits from current PR and update the 38baf32dd5d28bdd3ac22046716747147de975dd?

MarcSabatella commented:
Remaining commits now in a new PR #5270.
